cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -22,7 +22,12 @@ pub enum YaccParserErrorKind {
     IllegalName,
     IllegalString,
     IncompleteRule,
-    DuplicateRule,
+
+    DuplicateRule {
+        /// While [`YaccParserError`] has the span of the duplicate.<br/>
+        /// `original_span` contains a span for the rule being duplicated.
+        original_span: Span,
```

Comment:
I guess now, I wonder if perhaps this would be better as something like `original: DuplicatedItem{DuplicateRule, span}`.
which could allow us to implement an `Error` like `format`, and message?

Reply:
I'm not going to pretend I've followed that!

I'm broadly supportive of the general idea behind this PR *but* if we're going to break the API, let's make it more future proof. It should cope with multiple copies, so something like `DuplicateRules(Vec<Span>)` would be safer in the long run.

Reply:
Sorry about that, was attempting to sleep and write stuff down at the same time.

The general gist of it is, I was wondering if we want to provide a string message along with the span,
that mirrors the original "Duplicate rule" string but provides some context for the span.

like "The first instance of the duplicated rule", "additional duplicate rule",
we'll need some enum to match on, for implementing format strings, and it might make sense to just use the `newtype` pattern and use `YaccParserErrorKind` as exists in master for that.

So we could have something like `Vec<Newtype(YaccParserErrorKind, Span)>` in error where `newtype` provides:

```
// in the newtype
match self.kind {
   YaccParserErrorKind::DuplicateRule = f("the original rule duplicated"),
   _ => { "Does not have additional data" },
}
```

The `Other` variant with `YaccParserErrorRelatedDataKind`, was for fields which don't exactly correspond to one of the top-level error kinds, like "additional duplicate rule".

So basically I was thinking that perhaps we shouldn't add the span to kind, but put it along side a kind in the `YaccParserError`.

Part of the thinking is that this would be more future proof for errors we can't easily get spans for now, but possibly could
in the future, we can just leave an empty `Vec<(newtype, span)>` when we can't.

Reply:
Nope, I still don't understand :)

Reply:
No worries, trying to implementing it, will perhaps be better than trying to explain it, and it may end up a terrible idea.